### ISSUE TITLE
Use fixed parameter information for acquisition function evaluations when using`PartialFixedSampler`

### ISSUE THREAD

**Issue by nabenabe0928** [optuna/optuna#6620]
### Motivation

```python
import optuna


def objective(trial: optuna.Trial) -> float:
    x = trial.suggest_float("x", -5, 5)
    y = trial.suggest_float("y", -5, 5)
    return x**2 + y**2

base_sampler = optuna.samplers.GPSampler()
# or base_sampler = optuna.samplers.TPESampler(multivariate=True)
study = optuna.create_study(sampler=base_sampler)
study.optimize(objective, n_trials=20)
study.sampler = optuna.samplers.PartialFixedSampler(fixed_params={"x": study.best_params["x"]}, base_sampler=base_sampler)
study.optimize(objective, n_trials=10)

```

In the example above, the surrogate fitting should be performed along with the fixed parameters as well, but the fixed parameters are simply dropped during the training and the acquisition function optimization.
Since the fixed parameters also affect output prediction (and of course the acquisition function landscape as well), they should also be accounted.

### Description

1. Use the fixed parameters for surrogate fitting
2. Optimize unfixed parameters while fixing the fixed paraemters during the acquisition function optimization

This is very useful especially when we use Optuna for very expensive HPO, where we would like to tune, for example, learning rate first and tune the rest thereafter.

**Comment by SowmyaaDixit** [comment]
@nabenabe0928 : I pushed in a PR to solve this issue. Please take a look and review my changes when you get a chance. Thanks.

**Comment by greekera1000** [comment]
Hi @nabenabe0928, I’ve been looking into this issue and the previous attempts to fix it. I’m interested in digging into the `GPSampler` side of the problem and trying to address the fixed-parameter handling properly. Would it be okay if I worked on this and opened a PR?
i'll try my best to solve this

**Comment by nabenabe0928** [comment]
What's your plan?
Please describe it in 100--200 words.

@greekera1000

**Comment by greekera1000** [comment]
thankd @nabenabe0928  for the response ---Why the earlier attempts fell short, as I read it: enlarging the search space and overwriting the fixed keys afterward leaves those dimensions as free variables during acquisition optimization, so the selected free params are chosen under the wrong conditioning.

Looking at _gp/optim_mixed.py, the pattern needed already exists. _gradient_ascent_batched optimizes over continuous_indices only, while carrying the full parameter vector and writing the optimization variables back into those columns before evaluating the acquisition function. Optimization variables are already a subset of the full dims; they're just partitioned by continuous/discrete rather than free/fixed.

So the minimal design I'd propose centers on SearchSpace:

Include the fixed dimensions in the search space dict, so get_normalized_params reads their historical values and the GP fits on full vectors.
Make SearchSpace aware of fixed dims, then exclude those indices from continuous_indices / discrete_indices so local search cannot move them, and pin them to their normalized fixed value in sample_normalized_params so preliminary sampling is drawn on the fixed slice.

No reconstruction layer is needed, since full vectors are carried end-to-end. The cost is that SearchSpace becomes stateful about fixedness and the index arrays no longer mean "all continuous dims."

Two questions before I write code: should the fixed values reach the sampler via trial.system_attrs["fixed_params"] or an explicit argument? And should the initial scope be GPSampler only, or does TPESampler(multivariate=True) need a parallel design here?

**Comment by greekera1000** [comment]
Following up on my earlier comment with an implementation and benchmark results.
I'd like to propose a design for this, along with measurements from a working branch.

### Approach

The fixed dimensions are made visible to `GPSampler` rather than handled purely at the `PartialFixedSampler` layer:

- **Full-dimension GP fitting:** Fixed dims stay in the relative search space, so `get_normalized_params` reads their historical values and the GP is fit on full vectors.
- **Search space restriction:** `SearchSpace` takes an optional `fixed_params`. Fixed indices are removed from `continuous_indices` and `discrete_indices`, so local search structurally cannot move them, and they are pinned to their normalized value in `sample_normalized_params`, ensuring preliminary sampling is drawn on the fixed slice.
- **Slice projection:** Warmstart points passed to `_optimize_acqf` are projected onto the slice; otherwise, an off-slice best trial reintroduces the original problem.

The reason this implementation stays small is that the acquisition optimizer already updates only a subset of coordinates at a time while evaluating on the full vector. In `_gradient_ascent_batched`, the optimization variables are `continuous_indices`, written into a full-width array before `eval_acqf`; `_local_search_discrete` does the same for the discrete coordinates, and `local_search_mixed_batched` alternates between them. 

Those coordinates are normally only held constant within a single step — but because both index arrays are derived from `SearchSpace`, dropping the fixed indices from each means neither phase of the alternation ever selects them. Consequently, they stay at their prescribed values for the entire optimization while the acquisition function is still evaluated on the complete vector. That is why no separate free-dims to full-dims reconstruction layer is needed. 

The diff is about 50 lines across `_gp/search_space.py`, `samplers/_gp/sampler.py`, and `samplers/_partial_fixed.py`.

---

### Measurements

Using the objective from #6634, scoring only the trials sampled under `PartialFixedSampler`, evaluated against the best value attainable given the fixed parameter.

**Setup:** 100 paired runs: 20 seeds × 5 fixed-`y` settings (one of which is `y = best_params["y"]` as in the original report).

| Metric | `master` | This design |
| :--- | :--- | :--- |
| **Mean gap** | 0.431 | 0.004 |
| **Max gap** | 3.100 | 0.027 |
| **Runs with gap > 0.1** | 36 | 0 |

The branch is better on 78 of 100 paired runs and worse on 22, but its worst case across every run is 0.027 against `master`'s 3.100. A paired bootstrap on the mean difference yields **−0.427**, 95% CI [−0.587, −0.284].

---

### Caveats

1. This benchmark is currently a single 2-D objective with a single fixed float. I have not yet tested higher dimensions or fixed categoricals, though the categorical path via `to_internal_repr` is implemented.
2. I see an occasional `invalid value encountered in divide` at `optim_mixed.py:311` on this branch that does not appear on `master`, which I am actively investigating.

---

### Open Questions Before Opening a PR

1. **How should fixed values reach the sampler?** My branch currently special-cases `GPSampler` inside `PartialFixedSampler`, which I don't think is acceptable as a clean long-term design. Potential alternatives:
   - Reading from `trial.system_attrs`
   - Providing an explicit sampler-level interface / argument
2. **Scope beyond `GPSampler`:** Since this issue is not isolated to `PartialFixedSampler`, should `TPESampler(multivariate=True)` be covered under this same design discussion?
